x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/PersistentCache.java

```diff
@@ -0,0 +1,623 @@
+                        final Path shardCachePath = getShardCachePath(new ShardPath(false, shardDataPath, shardDataPath, shardId));
+
+                        if (Files.isDirectory(shardCachePath)) {
+                            logger.trace("found snapshot cache dir at [{}], loading cache files from disk and index", shardCachePath);
+                            Files.walkFileTree(shardCachePath, new CacheFileVisitor(cacheService, writer, documents));
+                        }
+                    }
+                }
+            }
+            for (CacheIndexWriter writer : writers) {
+                writer.prepareCommit();
+            }
+            for (CacheIndexWriter writer : writers) {
+                writer.commit();
+            }
+            logger.info("persistent cache index loaded");
+            documents.clear();
+        } catch (IOException e) {
+            try {
+                close();
+            } catch (Exception e2) {
+                logger.warn("failed to close persistent cache index", e2);
+                e.addSuppressed(e2);
+            }
+            throw new UncheckedIOException("Failed to load persistent cache", e);
+        } finally {
+            closeIfAnyIndexWriterHasTragedyOrIsClosed();
+        }
+    }
+
+    void commit() throws IOException {
+        ensureOpen();
+        try {
+            for (CacheIndexWriter writer : writers) {
+                writer.prepareCommit();
+            }
+            for (CacheIndexWriter writer : writers) {
+                writer.commit();
+            }
+        } catch (IOException e) {
+            try {
+                close();
+            } catch (Exception e2) {
+                logger.warn("failed to close persistent cache index writer", e2);
+                e.addSuppressed(e2);
+            }
+            throw e;
+        } finally {
+            closeIfAnyIndexWriterHasTragedyOrIsClosed();
+        }
+    }
+
+    private void closeIfAnyIndexWriterHasTragedyOrIsClosed() {
+        if (writers.stream().map(writer -> writer.indexWriter).anyMatch(iw -> iw.getTragicException() != null || iw.isOpen() == false)) {
+            try {
+                close();
+            } catch (Exception e) {
+                logger.warn("failed to close persistent cache index", e);
+            }
+        }
+    }
+
+    public boolean hasDeletions() {
+        ensureOpen();
+        for (CacheIndexWriter writer : writers) {
+            if (writer.indexWriter.hasDeletions()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public long getNumDocs() {
+        ensureOpen();
+        long count = 0L;
+        for (CacheIndexWriter writer : writers) {
+            count += writer.indexWriter.getPendingNumDocs();
+        }
+        return count;
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (closed.compareAndSet(false, true)) {
+            IOUtils.close(writers);
```

Comment:
NIT: Maybe just  `IOUtils.close(writers, documents::clear);` to make it clear no matter what? (not that it matters much)

Reply:
I don't think I can bake `documents::clear` into `IOUtils.close()` but I can definitely clear the documents map in a finally block. I pushed ac6050f206ff4e096736ef19a2812c33a365b9b7.

Reply:
:facepalm: right thanks, missed the plural on `writers` :)